src/io/iter_thread_csv-inl.hpp

```diff
@@ -0,0 +1,291 @@
+      if (has_header) {
+        char ch;
+        while ((ch = fgetc(fpcsv)) != EOF) {
+          if (ch == '\r' || ch == '\n') break;
+        }
+      }
+    }
+    inline bool Init() {
+      list_order.resize(path_csv.size());
+      for (size_t i = 0; i < path_csv.size(); ++i) {
+        list_order[i] = i;
+      }
+      if (shuffle != 0) {
+        rnd.Shuffle(list_order);
+      }
+      // load in data
+      list_ptr = 0;
+      fpcsv = utils::FopenCheck(path_csv[list_order[0]].c_str(), "r");
+      this->SkipHead();
+      utils::Check(this->FillBuffer(), "CSVIterator: first csv must be valid");
+      return true;
+    }
+    inline void SetParam(const char *name, const char *val) {
+      if (!strcmp(name, "label_width")) {
+        label_width = atoi(val);
+      }
+      if (!strcmp(name, "shuffle")) {
+        shuffle = atoi(val);
+      }
+      if (!strcmp(name, "seed_data")) {
+        rnd.Seed(atoi(val) + kRandMagic);
+      }
+      if(!strcmp(name, "has_header"))   has_header = atoi(val);
+      if (!strcmp(name, "input_shape")) {
+        utils::Check(sscanf(val, "%u,%u,%u", &shape[0], &shape[1], &shape[2]) == 3,
+                   "input_shape must be three consecutive integers without space example: 1,1,200 ");
+      }
+    }
+    inline bool FillBuffer() {
+      index_t idx = 0;
+      while (!feof(fpcsv)) {
+        if (idx >= entry.size()) {
+          entry.push_back(CacheEntry(label_width, shape));
+        }
+        if (fscanf(fpcsv, "%f", &entry[idx].label[0]) != EOF) {
+          for (int i = 1; i < label_width; ++i) {
+            utils::Check(fscanf(fpcsv, ",%f", &entry[idx].label[i]) == 1,
+                        "CSVIterator: Error when reading label. Possible incorrect file or label_width.");
+          }
+          for (index_t i = 0; i < shape[0]; ++i) {
+            for (index_t j = 0; j < shape[1]; ++j) {
+              for (index_t k = 0; k < shape[2]; ++k) {
+                 utils::Check(fscanf(fpcsv, ",%f", &entry[idx].data[i][j][k]) == 1,
+                              "CSVIterator: Error when reading data. Possible incorrect file or input_shape.");
+              }
+            }
+          }
+        }
+        idx++;
+      }
+      inst_order.resize(std::max(entry.size(), inst_order.size()));
+      for (index_t i = 0; i < inst_order.size(); ++i) {
+        inst_order[i] = i;
+      }
+      if (shuffle) {
+        rnd.Shuffle(inst_order);
+      }
+
+      list_ptr += 1;
+      data_ptr = 0;
+      return true;
+    }
+    inline DataInst Create(void) {
+      DataInst a;
+      a.data.shape_ = shape;
+      a.label.shape_ = mshadow::Shape1(label_width);
+      mshadow::AllocSpace(&a.data, false);
+      mshadow::AllocSpace(&a.label, false);
+      return a;
+    }
+    inline void FreeSpace(DataInst &a) {
+      mshadow::FreeSpace(&a.data);
+      mshadow::FreeSpace(&a.label);
+    }
+    inline bool LoadNext(DataInst &val) {
+      while (true) {
+        if (data_ptr >= entry.size()) {
```

Comment:
should be page->Size(), entry.size() can be bigger than content of current result